### DISCUSSION TITLE
Capacity and proof of concept compared to EES

### DISCUSSION THREAD

**Discussion by max3-2** [sympy/sympy#23263] [Q&A]
Hello,

currently, we are using EES to simulate thermodynamic cycles. The modeling there is, in my opinion, very compiclated and tedious to expand or script. So as a python user for many years, I thought about `sympy`. Before I spent the time with a demonstrator, I thought to get some input from experienced `sympy` users - while I know the concept and have used it for smaller projects, I am in no position to judge the following:

- The problems usually consists of thermodynamic cycles with refrigerants, most of the equations are states (which I plan to pull from `CoolProp`) and linear equations. Some non-linearities exists as far as I can see and there is the possibility of using integral and differential equations in the future.
- The user inputs some boundary conditions, some are pulled from component tables, some are interpolated. All rely on the input quantities. The component equations are usually numeric, e.g. yielding definite number values for symbolic quantities, however *some rely on the symbolic computations*
- I would pin the number of variables around a few hundred, including all state vars.
- The problem usually is defined or it will be solved using input tables for multiple vars, which is still well defined.

So can you help me out if there is any use to start the road down using `sympy` - maybe even someone with experience with `EES`? There seems to be some iterations going on during solving even the system is defined - this is for stability I assume. Also, the order of entered equations does not matter, they are collected, compiled and solved together.
I do see the biggest issues in the order of execution and the subroutines where symbolic and numeric mix. Is this possible with `sympy` - I recall being able to replace some quantities in equations.

**Comment by oscarbenjamin** [comment]
It would help if you could give a concrete (simplified) example of something that you want to do e.g. what are the equations and what do you want to do with them?

**Comment by max3-2** [comment]
Well since even simple processes depend on quite a lot this will take some time, and I wanted to first evaluate the possibilities before spending said time. I ill try to outline a typical system below an hope that helps. If not, I will need more time in advance...

In general, the cycle is defined by states along it, in a very simple case, 4. Imaging a fridge. Depending on the input the aim is to calculate some defining variables. An example would be, where h_x are states from temperatures and pressures and I hope the rest is fairly clear (heat fluxes, mass fluxes, pressures, temperatures)

```python
Q_net = m_dot * (h3 -h2)  # Net heat removal, you would specify this value
Q_in = m_dot * (h4 - h1)  # Added heat
h3 = h4  # That's a throttle
P = m_dot * (h2 - h1)
Q_net = Q_in + P

# Unknowns: 8
# Equations: 5
# Targets: Q_net

# Need two more quantities
h4 = 'specified via Temperature / Pressure'
h1 = 'specified via delta from h4'

# Add code to solve for Q_net. This is easy in this case but assume more equations and states in between, see below
```

The example above will increase in complexity by adding non-ideal parts (pressure drops, heat losses) or more states (two phase flow, other media). Especially the compression and the two heat exchanges will be most likely defined by more complex equations considering the state (user input pressure and temperatures) and some geometric configurations (which are defined before and thus replace symbols in the corresponding equations *during* setup). Does that help?

**Reply by oscarbenjamin** [reply]
I think if you leave out the physics and just explain what you want to do with the equations then it would be easier to answer. It sounds as if you want to solve a system of possibly algebraic or possibly differential equations possibly analytically, possibly numerically, possibly after substituting some values. All of those things are doable.

In your example I suppose that you want to solve a system of algebraic equations. This case is simple enough to be symbolically solvable:
```python
In [97]: P, Q_net, Q_in, m_dot, h1, h2, h3, h4 = symbols('P, Q_net, Q_in, m_dot, h1:5')

In [98]: eqs = [
    ...:     Eq(Q_net, m_dot * (h3 -h2)),
    ...:     Eq(Q_in, m_dot * (h4 - h1)),
    ...:     Eq(h3, h4),
    ...:     Eq(P, m_dot * (h2 - h1)),
    ...:     Eq(Q_net, Q_in + P),
    ...: ]

In [100]: sol1, sol2 = solve(eqs, [Q_net, Q_in, m_dot, h2, h3, P], dict=True)

In [101]: sol1
Out[101]: {P: 0, Qᵢₙ: 0, Qₙₑₜ: 0, h₃: h₄, m_dot: 0}

In [102]: sol2
Out[102]: {P: 0, Qᵢₙ: -m_dot⋅(h₁ - h₄), Qₙₑₜ: -m_dot⋅(h₁ - h₄), h₂: h₁, h₃: h₄}
```
The first solution set has `Qnet = 0`.  The second has `Qnet = -m_dot*(h1-h4)` but `m_dot` can be anything.

**Comment by max3-2** [comment]
First of all thank you for your example. I am currently using the little time I have to look further into it but need some more time. To roughly answer the current questions:
- It will be a mix of equations, but for a start most of the are analytical and fairly simple like the ones in my example. There might be some diff / int added but this is optional. Obviously then the diffs / integrals would need to be defined.
- The aim is to solve for some numerical quantities. If the way is to solve the system and then substitute or the way around - I do not know however I feel the symbolic solution and then substitute is faster especially when running parameter studies and then being able to substitute with arrays

Two main question are still open, maybe you can provide some input there. I will do my best to build a demonstrator but this will take me time (time in weeks...)

1. **How does it scale?** Assuming an equations style as above, and a defined system with unknowns == equations, but having hundreds of equations. Is this predictable in terms of computational time?

2. Regarding your example, I further simplified an example. See below. I count 7 symbols and 7 equations. Can you elaborate on:

- There does not seem to be a numerical solution, `h4` is not replaced. Why?
- Assuming this works and yields a single numeric value, I would like to leave the last specification out, solve for `P = f(Q_out)` and then run this using an array for `Q_out` - possible?

```python
import sympy as sp
from sympy import Eq

P, Q_out, Q_in, h1, h2, h3, h4 = sp.symbols('P, Q_out, Q_in, h1:5')

eqs = [
    Eq(Q_out, h3 - h2),
    Eq(Q_in, h1 - h4),
    Eq(h3, h4),
    Eq(P, h2 - h1),
    Eq(Q_out + Q_in + P, 0),
    ]

specifications = [
    Eq(h1, 550.),
    Eq(Q_out, -300),
    ]

sols = sp.solve(eqs + specifications, [P, Q_out, Q_in, h1, h2, h3, h4],
                dict=True)
print(sols)

--> [{P: h4 - 250.0, Q_out: -300.000000000000, Q_in: 550.0 - h4, h1: 550.000000000000, h2: h4 + 300.0, h3: h4}]

**Reply by oscarbenjamin** [reply]
> * There does not seem to be a numerical solution, `h4` is not replaced. Why?

The equations are (`for eq in eqs + specifications: pprint(eq)`):
```python
Qₒᵤₜ = -h₂ + h₃
Qᵢₙ = h₁ - h₄
h₃ = h₄
P = -h₁ + h₂
P + Qᵢₙ + Qₒᵤₜ = 0
h₁ = 550.0
Qₒᵤₜ = -300
```
The solution found is parametric (given in terms of `h4`) because the system of equations does not have a unique solution. Since this example is linear you can see this by calculating the echelon form of the augmented matrix:
```python
In [12]: M, b = linear_eq_to_matrix(eqs + specifications, [P, Q_out, Q_in, h1, h2, h3, h4])

In [13]: Maug = Matrix.hstack(M, b)

In [14]: Maug
Out[14]: 
⎡0  1  0  0   1   -1  0     0  ⎤
⎢                              ⎥
⎢0  0  1  -1  0   0   1     0  ⎥
⎢                              ⎥
⎢0  0  0  0   0   1   -1    0  ⎥
⎢                              ⎥
⎢1  0  0  1   -1  0   0     0  ⎥
⎢                              ⎥
⎢1  1  1  0   0   0   0     0  ⎥
⎢                              ⎥
⎢0  0  0  1   0   0   0   550.0⎥
⎢                              ⎥
⎣0  1  0  0   0   0   0   -300 ⎦

In [15]: Maug.echelon_form()
Out[15]: 
⎡1  0  0  1   -1  0   0     0  ⎤
⎢                              ⎥
⎢0  1  0  0   1   -1  0     0  ⎥
⎢                              ⎥
⎢0  0  1  -1  0   0   1     0  ⎥
⎢                              ⎥
⎢0  0  0  1   0   0   0   550.0⎥
⎢                              ⎥
⎢0  0  0  0   -1  1   0   -300 ⎥
⎢                              ⎥
⎢0  0  0  0   0   1   -1    0  ⎥
⎢                              ⎥
⎣0  0  0  0   0   0   0     0  ⎦
```
Note the row of all zeros at the bottom. You might have 7 equations for 7 unknowns but they are linearly dependent and so the matrix is not invrtible and you don't get a unique solution:
```python
In [16]: M.det()
Out[16]: 0
```

> How does it scale? Assuming an equations style as above, and a defined system with unknowns == equations, but having hundreds of equations. Is this predictable in terms of computational time?

The answer to this depends very strongly on various things that I can't really know. The example you've just shown is linear with all explicit numeric coefficients. That will scale fine to 1000s of equations although some tricks might be needed to make it faster than solve.

On the other hand when there are polynomials or symbolic parameters then depending on the structure of the system even a small system can be extremely expensive to solve analytically. If the equations are as simple as the examples you've shown it should be fine but it can be surprising sometimes how much a small change in the system of equations affects (analytic) solvability. If there are well-defined numeric solutions and no symbolic parameters then it should usually be possible to get them with a numeric solver though.
